// pcf-scripts/constants.js
"use strict";
// Copyright (C) Microsoft Corporation. All rights reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.BUILD_TOOL_NAME = 'pac';
exports.BUNDLE_NAME = 'bundle.js';
exports.CODE_ELEM_NAME = 'code';
exports.COLOR_RED_RESET = '\x1b[31m%s\x1b[0m';
exports.COMPONENT_FRAMEWORK_PACKAGE_NAME = '@types/powerapps-component-framework';
exports.CONFIGURATION_FILE_NAME = 'pcfconfig.json';
exports.DEFAULT_BUILD_MODE = 'development';
exports.DEFAULT_LANGUAGE = 'en';
exports.DEFAULT_OUT_DIR = 'out/controls';
exports.LIBRARY_ELEM_NAME = 'library';
exports.MANIFEST_INPUT_FILE_NAME = 'ControlManifest.Input.xml';
exports.MANIFEST_TYPE_FILE_NAME = 'ManifestTypes.d.ts';
exports.PACKAGE_LOCK_JSON_PATH = 'package-lock.json';
exports.PACKAGE_JSON_PATH = 'package.json';
exports.PCF_SCRIPTS_PACKAGE_NAME = 'pcf-scripts';
exports.PROCESSED_MANIFEST_FILE_NAME = 'ControlManifest.xml';
exports.TEMP_NAMESPACE = 'pcf_tools_652ac3f36e1e4bca82eb3c1dc44e6fad';
exports.TYPING_FOLDER_NAME = 'generated';
exports.NODE_MODULES = 'node_modules';
