// pcf-start/diagnosticMessages.generated.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function diag(code, category, key, message) {
    return { code: code, category: category, key: key, message: message };
}
/* tslint:disable:max-line-length */
/* tslint:disable:quotemark */
exports.strings = {
    loading_harness: diag(1002, "Info", "loading_harness", "Loading control harness..."),
    loading_control: diag(1003, "Info", "loading_control", "Loading control..."),
    inputs: diag(1004, "Info", "inputs", "Inputs"),
    property: diag(1005, "Info", "property", "Property"),
    value: diag(1006, "Info", "value", "Value"),
    type: diag(1007, "Info", "type", "Type"),
    error_parse_resource_string: diag(1008, "Error", "error_parse_resource_string", "Error parsing resource '{0}'"),
    outputs: diag(1009, "Info", "outputs", "Outputs"),
    resource_file_loaded: diag(1010, "Error", "resource_file_loaded", "Resource file '{0}' loaded."),
    error_resource_load: diag(1011, "Error", "error_resource_load", "Failed to load resource '{0}'"),
    error_resource_type_not_supported: diag(1012, "Error", "error_resource_type_not_supported", "Resource '{0}' not loaded because {1} is not supported."),
    unknown_property_usage: diag(1013, "Error", "unknown_property_usage", "Property has an unknown value for the usage attribute"),
    unrecognized_type_group: diag(1014, "Error", "unrecognized_type_group", "Property references an unknown type-group '{0}'"),
    property_missing_type: diag(1015, "Error", "property_missing_type", "Property has neither type-of nor type-group attribute"),
    error_parsing_control_manifest: diag(1016, "Error", "error_parsing_control_manifest", "Manifest Parsing Error: Cannot parse ControlManifest.xml."),
    error_http_get_resource: diag(1017, "Error", "error_http_get_resource", "GET resource '{0}' returned with status {1}"),
    invoke_method: diag(1018, "Info", "invoke_method", "Invoked method {0} on {1} interface."),
    function_parameters: diag(1019, "Info", "function_parameters", "Parameters"),
    message: diag(1020, "Info", "message", "Message"),
    details: diag(1021, "Info", "details", "Details"),
    error_code: diag(1022, "Info", "error_code", "Error Code")
};
